File: imu_calib_matlab/src/runtime/get_bias_from_temperature.m
function [biasT, info] = get_bias_from_temperature(temp, biasConst, tempModel, varargin)
%GET_BIAS_FROM_TEMPERATURE Evaluate a temperature-dependent 3-axis bias model.
% Usage:
%   [biasT, info] = get_bias_from_temperature(temp, biasConst, tempModel)

opts = parse_options(varargin{:});
biasConst = ensure_vector3(biasConst, 'biasConst').';

if nargin < 1 || isempty(temp)
    biasT = biasConst;
    info = struct('valid', false, ...
                  'used_temperature_model', false, ...
                  'out_of_range', false, ...
                  'message', sprintf('Temperature vector not provided. Falling back to constant %s.', opts.target_name));
    return;
end

temp = double(temp(:));
if any(~isfinite(temp))
    error('get_bias_from_temperature:InvalidTemperature', ...
        'temp must contain finite numeric values.');
end

N = numel(temp);
biasT = repmat(biasConst, N, 1);
info = struct('valid', false, ...
              'used_temperature_model', false, ...
              'out_of_range', false, ...
              'message', sprintf('No valid temperature model. Falling back to constant %s.', opts.target_name));

if nargin < 3 || isempty(tempModel) || ~isstruct(tempModel)
    return;
end
if ~isfield(tempModel, 'valid') || ~tempModel.valid
    if isfield(tempModel, 'message') && ~isempty(tempModel.message)
        info.message = char(string(tempModel.message));
    end
    return;
end

[tempEval, rangeInfo] = prepare_temperature_input(temp, tempModel);
modelType = '';
if isfield(tempModel, 'type') && ~isempty(tempModel.type)
    modelType = lower(char(string(tempModel.type)));
end

switch modelType
    case 'poly'
        coeffs = [];
        if isfield(tempModel, 'coeffs') && ~isempty(tempModel.coeffs)
            coeffs = tempModel.coeffs;
        elseif isfield(tempModel, 'coeff') && isstruct(tempModel.coeff) && ...
                isfield(tempModel.coeff, 'poly_coeff')
            coeffs = tempModel.coeff.poly_coeff;
        end
        coeffs = double(coeffs);
        if size(coeffs, 2) ~= 3
            info.message = sprintf( ...
                'Temperature model coeffs must have shape [P x 3]. Falling back to constant %s.', opts.target_name);
            return;
        end
        referenceTemperature = getfield_or_default(tempModel, 'reference_temperature', 0.0);
        dT = tempEval - referenceTemperature;
        for axisIdx = 1:3
            biasT(:, axisIdx) = polyval(coeffs(:, axisIdx).', dT);
        end

    case 'piecewise_linear'
        coeff = getfield_or_default(tempModel, 'coeff', struct());
        breakpoints = getfield_or_default(tempModel, 'breakpoints', getfield_or_default(coeff, 'breakpoints', []));
        values = getfield_or_default(tempModel, 'values', getfield_or_default(coeff, 'values', []));
        breakpoints = double(breakpoints(:));
        values = double(values);
        if size(values, 2) ~= 3 || size(values, 1) ~= numel(breakpoints)
            info.message = sprintf( ...
                'Piecewise temperature model size mismatch. Falling back to constant %s.', opts.target_name);
            return;
        end
        for axisIdx = 1:3
            biasT(:, axisIdx) = interp_with_extrapolation(tempEval, breakpoints, values(:, axisIdx), tempModel);
        end

    otherwise
        info.message = sprintf( ...
            'Unsupported temperature model type "%s". Falling back to constant %s.', ...
            char(string(getfield_or_default(tempModel, 'type', ''))), opts.target_name);
        return;
end

info.valid = true;
info.used_temperature_model = true;
info.out_of_range = rangeInfo.out_of_range;
if isempty(rangeInfo.message)
    info.message = sprintf('Temperature-dependent %s(T) evaluated successfully.', opts.target_name);
else
    info.message = rangeInfo.message;
end
end

function opts = parse_options(varargin)
opts = struct('target_name', 'bias');
if mod(numel(varargin), 2) ~= 0
    error('get_bias_from_temperature:InvalidInput', ...
        'Optional arguments must be provided as name/value pairs.');
end
for k = 1:2:numel(varargin)
    name = lower(char(varargin{k}));
    value = varargin{k + 1};
    switch name
        case 'target_name'
            opts.target_name = char(string(value));
        otherwise
            error('get_bias_from_temperature:UnknownOption', ...
                'Unknown option "%s".', char(varargin{k}));
    end
end
end

function [tempEval, info] = prepare_temperature_input(temp, tempModel)
tempEval = temp;
info = struct('out_of_range', false, 'message', '');
tempRange = getfield_or_default(tempModel, 'temperature_range', [NaN; NaN]);
tempRange = double(tempRange(:));
if numel(tempRange) ~= 2 || any(~isfinite(tempRange))
    return;
end

tmin = tempRange(1);
tmax = tempRange(2);
outOfRange = any(temp < tmin | temp > tmax);
if ~outOfRange
    return;
end

mode = lower(char(string(getfield_or_default(tempModel, 'extrapolation_mode', 'warn_and_clamp'))));
switch mode
    case 'clamp'
        tempEval = min(max(temp, tmin), tmax);
        info.out_of_range = true;
        info.message = 'Input temperature was clamped to the fitted range.';
    case 'warn_and_clamp'
        tempEval = min(max(temp, tmin), tmax);
        info.out_of_range = true;
        info.message = 'Input temperature exceeded the fitted range; values were clamped.';
    case 'extrapolate'
        info.out_of_range = true;
        info.message = 'Input temperature exceeded the fitted range; model extrapolation was used.';
    otherwise
        tempEval = min(max(temp, tmin), tmax);
        info.out_of_range = true;
        info.message = sprintf('Unknown extrapolation mode "%s"; values were clamped.', mode);
end
end

function values = interp_with_extrapolation(x, xp, fp, tempModel)
mode = lower(char(string(getfield_or_default(tempModel, 'extrapolation_mode', 'warn_and_clamp'))));
if numel(xp) < 2
    values = interp1(xp, fp, x, 'linear', 'extrap');
    return;
end

if ismember(mode, {'clamp', 'warn_and_clamp'})
    values = interp1(xp, fp, x, 'linear', 'extrap');
    return;
end

values = interp1(xp, fp, x, 'linear', 'extrap');
end

function value = getfield_or_default(S, fieldName, defaultValue)
if isstruct(S) && isfield(S, fieldName)
    value = S.(fieldName);
else
    value = defaultValue;
end
end

function v = ensure_vector3(v, name)
v = double(v(:));
if numel(v) ~= 3 || any(~isfinite(v))
    error('get_bias_from_temperature:InvalidVector', ...
        '%s must be a finite 3x1 vector.', name);
end
end

File: imu_calib_python/imu_calib/runtime/get_bias_from_temperature.py
from __future__ import annotations

import numpy as np

from imu_calib.utils.math_utils import as_vector3


def get_bias_from_temperature(
    temp: np.ndarray | None,
    bias_const: np.ndarray,
    temp_model: dict | None = None,
    *,
    target_name: str = "bias",
) -> tuple[np.ndarray, dict]:
    """Evaluate a temperature-dependent 3-axis bias model with fallback."""

    bias_const = as_vector3(bias_const, "bias_const")
    if temp is None or np.asarray(temp).size == 0:
        return bias_const[None, :], {
            "valid": False,
            "used_temperature_model": False,
            "out_of_range": False,
            "message": f"Temperature vector not provided. Falling back to constant {target_name}.",
        }

    temp = np.asarray(temp, dtype=float).reshape(-1)
    N = temp.size
    bias_T = np.tile(bias_const, (N, 1))
    info = {
        "valid": False,
        "used_temperature_model": False,
        "out_of_range": False,
        "message": f"No valid temperature model. Falling back to constant {target_name}.",
    }
    if not temp_model or not isinstance(temp_model, dict):
        return bias_T, info
    if not temp_model.get("valid", False):
        if "message" in temp_model:
            info["message"] = str(temp_model.get("message", info["message"]))
        return bias_T, info

    temp_eval, range_info = _prepare_temperature_input(temp, temp_model)
    model_type = str(temp_model.get("type", "")).lower()
    if model_type == "poly":
        coeffs = np.asarray(
            temp_model.get("coeffs", temp_model.get("coeff", {}).get("poly_coeff", [])),
            dtype=float,
        )
        if coeffs.ndim != 2 or coeffs.shape[1] != 3:
            info["message"] = f"Temperature model coeffs must have shape (P, 3). Falling back to constant {target_name}."
            return bias_T, info
        dT = temp_eval - float(temp_model.get("reference_temperature", 0.0))
        bias_T = np.column_stack([np.polyval(coeffs[:, axis], dT) for axis in range(3)])
    elif model_type == "piecewise_linear":
        coeff = temp_model.get("coeff", {})
        breakpoints = np.asarray(temp_model.get("breakpoints", coeff.get("breakpoints", [])), dtype=float).reshape(-1)
        values = np.asarray(temp_model.get("values", coeff.get("values", [])), dtype=float)
        if values.shape != (breakpoints.size, 3):
            info["message"] = (
                f"Piecewise temperature model size mismatch. Falling back to constant {target_name}."
            )
            return bias_T, info
        bias_T = np.column_stack(
            [_interp_with_extrapolation(temp_eval, breakpoints, values[:, axis], temp_model) for axis in range(3)]
        )
    else:
        info["message"] = (
            f'Unsupported temperature model type "{temp_model.get("type", "")}". Falling back to constant {target_name}.'
        )
        return bias_T, info

    info["valid"] = True
    info["used_temperature_model"] = True
    info["out_of_range"] = bool(range_info["out_of_range"])
    info["message"] = range_info["message"] or f"Temperature-dependent {target_name}(T) evaluated successfully."
    return bias_T, info


def _prepare_temperature_input(temp: np.ndarray, temp_model: dict) -> tuple[np.ndarray, dict]:
    temp = np.asarray(temp, dtype=float).reshape(-1)
    temp_range = temp_model.get("temperature_range", [np.nan, np.nan])
    extrapolation_mode = str(temp_model.get("extrapolation_mode", "warn_and_clamp")).lower()
    if len(temp_range) != 2 or not np.all(np.isfinite(temp_range)):
        return temp, {"out_of_range": False, "message": ""}

    tmin = float(temp_range[0])
    tmax = float(temp_range[1])
    out_of_range = bool(np.any((temp < tmin) | (temp > tmax)))
    if not out_of_range:
        return temp, {"out_of_range": False, "message": ""}

    if extrapolation_mode == "clamp":
        return np.clip(temp, tmin, tmax), {"out_of_range": True, "message": "Input temperature was clamped to the fitted range."}
    if extrapolation_mode == "warn_and_clamp":
        return np.clip(temp, tmin, tmax), {
            "out_of_range": True,
            "message": "Input temperature exceeded the fitted range; values were clamped.",
        }
    if extrapolation_mode == "extrapolate":
        return temp, {"out_of_range": True, "message": "Input temperature exceeded the fitted range; model extrapolation was used."}
    return np.clip(temp, tmin, tmax), {
        "out_of_range": True,
        "message": f'Unknown extrapolation mode "{extrapolation_mode}"; values were clamped.',
    }


def _interp_with_extrapolation(x: np.ndarray, xp: np.ndarray, fp: np.ndarray, temp_model: dict) -> np.ndarray:
    mode = str(temp_model.get("extrapolation_mode", "warn_and_clamp")).lower()
    values = np.interp(x, xp, fp)
    if mode in {"clamp", "warn_and_clamp"} or xp.size < 2:
        return values

    slope_lo = (fp[1] - fp[0]) / (xp[1] - xp[0])
    slope_hi = (fp[-1] - fp[-2]) / (xp[-1] - xp[-2])
    below = x < xp[0]
    above = x > xp[-1]
    if np.any(below):
        values[below] = fp[0] + slope_lo * (x[below] - xp[0])
    if np.any(above):
        values[above] = fp[-1] + slope_hi * (x[above] - xp[-1])
    return values
